### PR TITLE
Implement group based authorization for openLDAP

### PR DESCRIPTION
**Pull request by sanjay990**
Once the user's password is validated, the user's credentials will be
used to check if the user belongs to a particular group. The server
property 'authentication.ldap.group-dn' must have the group's
distinguished name.

This commit handles group check for OpenLDAP.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapServerConfig.java [hunk 1]
```diff
@@ -20,10 +20,10 @@
     private String ldapUrl;
     private String baseDistinguishedName;
     private boolean authenticationEnabled;
-    private String serverType = ServerType.ACTIVE_DIRECTORY.name();
+    private String serverType;
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapServerConfig.java:23]: I would still keep these default values, so that the user doesnt have to set these for AD

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java [hunk 2]
```diff
@@ -13,32 +13,60 @@
...
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
 public class OpenLdapBinder
         implements LdapBinder
 {
-    private final String baseDistinguishedName;
+    private static final Logger LOG = Logger.get(ActiveDirectoryBinder.class);
+    private final String userObjectClass;
+    private final Optional<String> baseDistinguishedName;
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: why is this optional? baseDistinguishedName is a mandatory parameter for password verification

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java [hunk 3]
```diff
@@ -13,32 +13,60 @@
...
     @Override
     public boolean checkForGroupMembership(String user, String groupDistinguishedName, DirContext context)
     {
-        throw new UnsupportedOperationException("Group membership based authorization not yet supported");
+        checkState(baseDistinguishedName.isPresent(), "Base distinguished name (DN) for user %s is missing", user);
+        SearchControls searchControls = new SearchControls();
+        searchControls.setSearchScope(SearchControls.SUBTREE_SCOPE);
+        String searchBase = baseDistinguishedName.get();
+        try {
+            String searchFilter = format("(&(objectClass=%s)(uid=%s)(memberof=%s))", userObjectClass, user, groupDistinguishedName);
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: iirc, memberof feature is turned off in openLDAP by default. Did you check if this is the case? Are you planning to add this a requirement that, for this code to work the openLDAP server should have the `memberof` attribute enabled? See this http://stackoverflow.com/questions/20397852/inetorgperson-with-member-or-memberof

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties [hunk 4]
```diff
@@ -23,7 +23,9 @@ http-server.https.enabled=true
 http-server.https.keystore.path=/etc/ldap/coordinator.jks
 http-server.https.keystore.key=testldap
 authentication.ldap.enabled=true
-authentication.ldap.url=ldaps://ldap-server:636
-authentication.ldap.ad-domain=presto.testldap.com
-authentication.ldap.base-dn=OU=Asia,DC=presto,DC=testldap,DC=com
-authentication.ldap.group-dn=CN=FinanceSubGroup,OU=America,DC=presto,DC=testldap,DC=com
+
+authentication.ldap.server-type=OPENLDAP
```
Comment (anusudarsan) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: please revert this. we can wait till we have the openldap container

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml [hunk 5]
```diff
@@ -7,6 +7,7 @@ databases:
     cli_ldap_authentication: true
     cli_ldap_truststore_path: /etc/ldap/cacerts.jks
     cli_ldap_truststore_password: testldap
-    cli_ldap_user_name: FinanceSubGroupUserInAsia
+    cli_ldap_user_name: ldapuser
```
Comment (anusudarsan) [inline presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml]: revert this too

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java (hunk 3 above)
Comment (sanjay990) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: yes, by default its disabled. We can mention that as a requirement. Also since we are running the tests on docker-containers, we can have that installed on docker.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapServerConfig.java (hunk 1 above)
Comment (sanjay990) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapServerConfig.java:23]: Default should be AD or openLDAP ?

Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapServerConfig.java:23]: AD, ditto for userObjectClass. keep it as 'person'

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties (hunk 4 above)
Comment (sanjay990) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: ok

File: presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml (hunk 5 above)
Comment (sanjay990) [inline presto-product-tests/conf/tempto/tempto-configuration-for-docker-ldap.yaml]: ok

File: presto-main/src/main/java/com/facebook/presto/server/security/ServerSecurityModule.java [hunk 6]
```diff
@@ -47,8 +47,9 @@ protected void setup(Binder binder)
                     .to(LdapFilter.class)
                     .in(Scopes.SINGLETON);
             if (ldapServerConfig.getServerType().equalsIgnoreCase(OPENLDAP.name())) {
-                checkState(ldapServerConfig.getBaseDistinguishedName() != null, "Missing property 'authentication.ldap.base-dn'");
-                checkState(ldapServerConfig.getGroupDistinguishedName() == null, "Group membership based authorization not yet supported");
+                if (ldapServerConfig.getGroupDistinguishedName() != null) {
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/ServerSecurityModule.java]: this if check is not needed. baseDN is a mandatory property for openldap

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java [hunk 7]
```diff
@@ -13,32 +13,60 @@
...
     {
-        return format("uid=%s,%s", user, baseDistinguishedName);
+        return format("uid=%s,%s", user, baseDistinguishedName.get());
     }
 
     @Override
     public boolean checkForGroupMembership(String user, String groupDistinguishedName, DirContext context)
     {
-        throw new UnsupportedOperationException("Group membership based authorization not yet supported");
+        checkState(baseDistinguishedName.isPresent(), "Base distinguished name (DN) for user %s is missing", user);
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: looks like this is the same code as the check in AD? Do you think we can move it to the LdapFilter?

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties [hunk 8]
```diff
@@ -23,7 +23,9 @@ http-server.https.enabled=true
...
+authentication.ldap.user-object-class=inetOrgPerson
+authentication.ldap.url=ldap://10.25.150.158:389
+authentication.ldap.base-dn=ou=Users,dc=10.25.150.158
+authentication.ldap.group-dn=cn=ldapgroup,ou=Users,dc=10.25.150.158
```
Comment (amrutagokhale) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: Is this IP address of OpenLDAP server temporary until we have OpenLDAP on docker?

Comment (amrutagokhale) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: Add a newline at the end of this file.

File: presto-main/src/main/java/com/facebook/presto/server/security/ActiveDirectoryBinder.java [hunk 9]
```diff
@@ -35,14 +26,13 @@
     private static final Logger LOG = Logger.get(ActiveDirectoryBinder.class);
     private final String activeDirectoryDomain;
     private final String userObjectClass;
-    private final Optional<String> baseDistinguishedName;
+    private final String searchInput = "sAMAccountName";
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/ActiveDirectoryBinder.java]: why do you have this as a field here, instead of just having it locally in the searchFilter method?

File: presto-main/src/main/java/com/facebook/presto/server/security/ActiveDirectoryBinder.java [hunk 10]
```diff
@@ -52,26 +42,8 @@ public String getBindDistinguishedName(String user)
     }
 
     @Override
-    public boolean checkForGroupMembership(String user, String groupDistinguishedName, DirContext context)
+    public String getSearchFilter(String user, String groupDistinguishedName)
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/ActiveDirectoryBinder.java]: maybe rename this method as groupCheckSearchFilter.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 11]
```diff
@@ -75,6 +80,12 @@ public LdapFilter(LdapServerConfig config, LdapBinder ldapBinder)
         }
     }
 
+    private static void sendChallenge(HttpServletResponse response)
+    {
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: please revert this refactoring. looks unrelated.

File: presto-main/src/main/java/com/facebook/presto/server/security/ActiveDirectoryBinder.java [hunk 12]
```diff
@@ -35,14 +26,13 @@
     private static final Logger LOG = Logger.get(ActiveDirectoryBinder.class);
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/ActiveDirectoryBinder.java]: LOG is unused

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java [hunk 13]
```diff
@@ -13,21 +13,26 @@
...
 import javax.inject.Inject;
-import javax.naming.directory.DirContext;
 
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
 public class OpenLdapBinder
         implements LdapBinder
 {
+    private static final Logger LOG = Logger.get(ActiveDirectoryBinder.class);
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: LOG is unused.

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java [hunk 14]
```diff
@@ -13,21 +13,26 @@
...
+    private final String userObjectClass;
     private final String baseDistinguishedName;
+    private final String searchInput = "uid";
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: ditto for searchInput to be local

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java [hunk 15]
```diff
@@ -13,21 +13,26 @@
...
 
     @Inject
     public OpenLdapBinder(LdapServerConfig config)
     {
-        baseDistinguishedName = requireNonNull(config.getBaseDistinguishedName(), "baseDistinguishedName is null");
+        userObjectClass = requireNonNull(config.getUserObjectClass(), "userObjectClass is null");
+        baseDistinguishedName = requireNonNull(config.getBaseDistinguishedName());
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: please revert this as it was before (line 30)

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties [hunk 16]
```diff
@@ -26,4 +26,4 @@ authentication.ldap.enabled=true
 authentication.ldap.url=ldaps://ldap-server:636
 authentication.ldap.ad-domain=presto.testldap.com
 authentication.ldap.base-dn=OU=Asia,DC=presto,DC=testldap,DC=com
```
Comment (anusudarsan) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: revert this file too

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 17]
```diff
@@ -53,19 +56,21 @@
 public class LdapFilter
         implements Filter
 {
-    private static final Logger LOG = Logger.get(LdapFilter.class);
     public static final String AUTHENTICATION_TYPE = "basic";
     public static final String LDAP_CONTEXT_FACTORY = "com.sun.jndi.ldap.LdapCtxFactory";
+    private static final Logger LOG = Logger.get(LdapFilter.class);
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: unrelated

File: presto-product-tests/conf/presto/etc/singlenode-ldap.properties [hunk 18]
```diff
@@ -26,4 +26,4 @@ authentication.ldap.enabled=true
...
-authentication.ldap.group-dn=CN=ParentGroup,OU=America,DC=presto,DC=testldap,DC=com
+authentication.ldap.group-dn=CN=ParentGroup,OU=America,DC=presto,DC=testldap,DC=com
```
Comment (amrutagokhale) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: You forgot to add a newline at the end of file.

Comment (sanjay990) [inline presto-product-tests/conf/presto/etc/singlenode-ldap.properties]: Added new line.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapBinder.java [hunk 19]
```diff
@@ -13,11 +13,9 @@
...
 package com.facebook.presto.server.security;
 
-import javax.naming.directory.DirContext;
-
 public interface LdapBinder
 {
     String getBindDistinguishedName(String user);
 
-    boolean checkForGroupMembership(String user, String groupDistinguishedName, DirContext context);
+    String groupCheckSearchFilter(String user, String groupDistinguishedName);
```
Comment (amrutagokhale) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapBinder.java]: perhaps rename it as `getGroupCheckSearchFilter`. Or `getGroupSearchFilter` or something similar. `groupCheckSearchFilter` doesn't convey what the function is doing. It might be perceived as "checking a search filter for a group"...

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 20]
```diff
@@ -119,7 +124,7 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
                 context = authenticate(environment);
 
                 if (groupDistinguishedName.isPresent()) {
-                    if (!ldapBinder.checkForGroupMembership(user, groupDistinguishedName.get(), context)) {
+                    if (!checkForGroupMembership(user, groupDistinguishedName.get(), context)) {
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: I would just move this throw statement to the method now, instead of checking the returned flag.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapBinder.java (hunk 19 above)
Comment (sanjay990) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapBinder.java]: I agree. I would like to go with getGroupSearchFilter
@anusudarsan : Any preferences.

**Comment by amrutagokhale** [comment]
nit: rewrite `openLDAP` as `OpenLDAP` in the commit message.

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapBinder.java (hunk 19 above)
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapBinder.java]: +1

**Comment by anusudarsan** [comment]
also update the commit message for things specific to this commit, like the requirement that `memberof` overlay must be enabled for OpenLDAP server for this check to work

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java [hunk 21]
```diff
@@ -184,6 +187,25 @@ public void destroy()
...
+        checkState(baseDistinguishedName.isPresent(), "Base distinguished name (DN) for user %s is missing", user);
+        SearchControls searchControls = new SearchControls();
+        searchControls.setSearchScope(SearchControls.SUBTREE_SCOPE);
+        String searchBase = baseDistinguishedName.get();
+        try {
+            String searchFilter = ldapBinder.getGroupSearchFilter(user, groupDistinguishedName);
+            LOG.debug("Group membership check for user '%s' using query: %s and base distinguished name: %s", user, searchFilter, searchBase);
+            NamingEnumeration<SearchResult> results = context.search(searchBase, searchFilter, searchControls);
+            if (!results.hasMoreElements()) {
+                throw new AuthenticationException(format("Authentication failed: Userssssssssbvdfds %s not a member of the group %s", user, groupDistinguishedName));
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapFilter.java]: typo: Userssssssssbvdfds

**Comment by anusudarsan** [comment]
Did you get AD tests passing? LGTM

**Comment by sanjay990** [comment]
yes . the tests are passing on AD

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java [hunk 22]
```diff
@@ -14,19 +14,20 @@
...
 public class OpenLdapBinder
         implements LdapBinder
 {
+    private final String userObjectClass;
     private final String baseDistinguishedName;
 
     @Inject
     public OpenLdapBinder(LdapServerConfig config)
     {
+        userObjectClass = requireNonNull(config.getUserObjectClass(), "userObjectClass is null");
```
Comment (amrutagokhale) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: `userObjectClass` will be `person` by default. So if someone forgets to set `authentication.ldap.user-object-class` property for OpenLDAP (which is possible since this property is not mandatory?), would this break the code? 
CC: @anusudarsan

Comment (sanjay990) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: yes, I think that's what we discussed. By default we have things for AD.

File: presto-main/src/main/java/com/facebook/presto/server/security/ServerSecurityModule.java [hunk 23]
```diff
@@ -46,9 +46,10 @@ protected void setup(Binder binder)
                     .addBinding()
                     .to(LdapFilter.class)
                     .in(Scopes.SINGLETON);
+            checkState(ldapServerConfig.getServerType() != null, "Missing property 'authentication.ldap.server-type'");
+            checkState(ldapServerConfig.getUserObjectClass() != null, "Missing property 'authentication.ldap.user-object-class'");
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/ServerSecurityModule.java]: hmm..the property user-object-class is mandatory only for group membership check, and not for password validation, right?. so maybe we shud check if group-dn is set first

**Comment by amrutagokhale** [comment]
Now that the two properties `serverType` and `userObjectClass` are mandatory, shouldn't you add those to default .properties file?

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapBinder.java [hunk 24]
```diff
@@ -17,5 +17,5 @@
 {
     String getBindDistinguishedName(String user);
 
-    String getGroupSearchFilter(String user, String groupDistinguishedName);
+    String getSearchInput();
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapBinder.java]: maybe call this getUserSearchInput

File: presto-main/src/main/java/com/facebook/presto/server/security/ActiveDirectoryBinder.java [hunk 25]
```diff
@@ -16,19 +16,16 @@
...
 {
     private final String activeDirectoryDomain;
-    private final String userObjectClass;
 
     @Inject
     public ActiveDirectoryBinder(LdapServerConfig config)
     {
-        activeDirectoryDomain = requireNonNull(config.getActiveDirectoryDomain(), "activeDirectoryDomain is null");
-        userObjectClass = requireNonNull(config.getUserObjectClass(), "userObjectClass is null");
+        activeDirectoryDomain = config.getActiveDirectoryDomain();
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/ActiveDirectoryBinder.java]: add requireNonNull

File: presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java [hunk 26]
```diff
@@ -16,19 +16,16 @@
...
 {
-    private final String userObjectClass;
     private final String baseDistinguishedName;
 
     @Inject
     public OpenLdapBinder(LdapServerConfig config)
     {
-        userObjectClass = requireNonNull(config.getUserObjectClass(), "userObjectClass is null");
-        baseDistinguishedName = requireNonNull(config.getBaseDistinguishedName(), "baseDistinguishedName is null");
+        baseDistinguishedName = config.getBaseDistinguishedName();
```
Comment (anusudarsan) [inline presto-main/src/main/java/com/facebook/presto/server/security/OpenLdapBinder.java]: requireNonNull

**Comment by anusudarsan** [comment]
some comments. otherwise LGTM

File: presto-main/src/main/java/com/facebook/presto/server/security/ServerSecurityModule.java [hunk 27]
```diff
@@ -46,24 +46,7 @@ protected void setup(Binder binder)
...
-                if (ldapServerConfig.getGroupDistinguishedName() != null) {
-                    checkState(ldapServerConfig.getBaseDistinguishedName() != null, "Missing property 'authentication.ldap.base-dn'");
-                }
-
-                binder.bind(LdapBinder.class).to(ActiveDirectoryBinder.class).in(Scopes.SINGLETON);
-            }
-            else {
-                throw new IllegalStateException(format("Invalid value '%s' for the property 'authentication.ldap.server-type'", ldapServerConfig.getServerType()));
-            }
+            ldapPropertyValidation(ldapServerConfig, binder);
```
Comment (amrutagokhale) [inline presto-main/src/main/java/com/facebook/presto/server/security/ServerSecurityModule.java]: A better name might be `setupLdap` (on similar lines of the `setup()` method)

**Comment by amrutagokhale** [comment]
LGTM % comment

**Comment by sanjay990** [comment]
merged with ldap-product-tests.
